src/components/DateTimePicker/DateTimePicker.jsx

```diff
@@ -0,0 +1,865 @@
+  presets: PRESET_VALUES,
+  intervals: [
+    {
+      label: 'minutes',
+      value: INTERVAL_VALUES.MINUTES,
+    },
+    {
+      label: 'hours',
+      value: INTERVAL_VALUES.HOURS,
+    },
+    {
+      label: 'days',
+      value: INTERVAL_VALUES.DAYS,
+    },
+    {
+      label: 'weeks',
+      value: INTERVAL_VALUES.WEEKS,
+    },
+    {
+      label: 'months',
+      value: INTERVAL_VALUES.MONTHS,
+    },
+    {
+      label: 'years',
+      value: INTERVAL_VALUES.YEARS,
+    },
+  ],
+  relatives: [
+    {
+      label: 'Yesterday',
+      value: RELATIVE_VALUES.YESTERDAY,
+    },
+    {
+      label: 'Today',
+      value: RELATIVE_VALUES.TODAY,
+    },
+    {
+      label: '',
+      value: '',
+    },
+  ],
+  expanded: false,
+  disabled: false,
+  showRelativeOption: true,
+  onCancel: null,
+  onApply: null,
+  i18n: {
+    toLabel: 'to',
+    toNowLabel: 'to Now',
+    calendarLabel: 'Calendar',
+    presetLabels: 'Last 30 minutes|Last 1 hour|Last 6 hours|Last 12 hours|Last 24 hours',
+    intervalLabels: 'minutes|hours|days|weeks|months|years',
+    relativeLabels: 'Yesterday|Today',
+    customRangeLinkLabel: 'Custom Range',
+    customRangeLabel: 'Custom range',
+    relativeLabel: 'Relative',
+    lastLabel: 'Last',
+    invalidNumberLabel: 'Number is not valid',
+    relativeToLabel: 'Relative to',
+    absoluteLabel: 'Absolute',
+    startTimeLabel: 'Start time',
+    endTimeLabel: 'End time',
+    applyBtnLabel: 'Apply',
+    cancelBtnLabel: 'Cancel',
+    backBtnLabel: 'Back',
+  },
+};
+
+const DateTimePicker = ({
+  defaultValue,
+  dateTimeMask,
+  presets,
+  intervals,
+  relatives,
+  expanded,
+  disabled,
+  showRelativeOption,
+  onCancel,
+  onApply,
+  i18n,
+  ...others
+}) => {
+  const strings = {
+    ...defaultProps.i18n,
+    ...i18n,
+  };
+
+  const [isExpanded, setIsExpanded] = useState(expanded);
+  const [customRangeKind, setCustomRangeKind] = useState(
+    showRelativeOption ? PICKER_KINDS.RELATIVE : PICKER_KINDS.ABSOLUTE
+  );
+
+  const [isCustomRange, setIsCustomRange] = useState(false);
+
+  const datePickerRef = React.createRef();
+
+  const [selectedPreset, setSelectedPreset] = useState(null);
+
+  const [currentValue, setCurrentValue] = useState(null);
+  const [lastAppliedValue, setLastAppliedValue] = useState(null);
+  const [humanValue, setHumanValue] = useState(null);
+
+  const [relativeValue, setRelativeValue] = useState(null);
+  const [absoluteValue, setAbsoluteValue] = useState(null);
+
+  const [focusOnFirstField, setFocusOnFirstField] = useState(true);
+
+  const dateTimePickerBaseValue = {
+    kind: '',
+    preset: {
+      label: null,
+      offset: null,
+    },
+    relative: {
+      lastNumber: null,
+      lastInterval: null,
+      relativeToWhen: null,
+      relativeToTime: null,
+    },
+    absolute: {
+      startDate: null,
+      startTime: null,
+      endDate: null,
+      endTime: null,
+    },
+  };
+
+  useEffect(() => {
+    window.setTimeout(() => {
```

Comment:
Doesn't this `setTimeout` need to be assigned to a variable and cleaned up via the useEffect return function to prevent infinite loops and memory leaks?

Reply:
imho in this case it's not needed, the dom reparenting or calendar opening do not trigger a rendering cycle on the component itself, so the effect is not called recursively by itself